pkg/operator/ceph/object/objectstore.go

```diff
@@ -88,78 +105,259 @@ func deleteRealmAndPools(context *Context, spec cephv1.ObjectStoreSpec) error {
+			return errors.Wrapf(err, "failed to find out whether zone group %q in is the master zone group", zoneGroupName)
 		}
+	}
+	_, err = RunAdminCommandNoRealm(context, "zone", "modify", realmArg, zoneGroupArg, zoneArg, endpointArg)
+	if err != nil {
+		return errors.Wrapf(err, "failed to add object store %q in rgw zone %q", context.Name, zoneName)
+	}
 
-		// create the zone if it doesn't exist yet
-		output, err = runAdminCommand(context, "zone", "get", zoneArg)
+	// the period will help notify other zones of changes if there are multi-zones
+	_, err = RunAdminCommandNoRealm(context, "period", "update", "--commit", realmArg, zoneGroupArg, zoneArg)
+	if err != nil {
+		return errors.Wrap(err, "failed to update period")
+	}
+	logger.Infof("added object store %q to realm %q, zonegroup %q, zone %q", context.Name, realmName, zoneGroupName, zoneName)
+
+	// create system user for realm for master zone in master zonegorup for multisite scenario
+	if zoneIsMaster && zoneGroupIsMaster {
+		uid := realmName + "-system-user"
+		uidArg := fmt.Sprintf("--uid=%s", uid)
+		output, err := RunAdminCommandNoRealm(context, "user", "info", uidArg)
 		if err != nil {
-			updatePeriod = true
-			output, err = runAdminCommand(context, "zone", "create", "--master", endpointArg, zoneArg)
-			if err != nil {
-				return errors.Wrapf(err, "failed to create rgw zonegroup for %q", context.Name)
+			if code, ok := exec.ExitStatus(err); ok && code == int(syscall.EINVAL) {
+				logger.Debugf("realm system user %q not found, running `radosgw-admin user create`", uid)
+				accessKeyArg, secretKeyArg, err := GetRealmKeyArgs(context.Context, realmName, namespace)
+				if err != nil {
+					return errors.Wrap(err, "failed to get keys for realm")
+				}
+				systemArg := "--system"
+				displayNameArg := fmt.Sprintf("--display-name=%s.user", realmName)
+				output, err = RunAdminCommandNoRealm(context, "user", "create", realmArg, zoneGroupArg, zoneArg, uidArg, displayNameArg, accessKeyArg, secretKeyArg, systemArg)
+				if err != nil {
+					return errors.Wrapf(err, "failed to create realm system user %q for reason: %q", uid, output)
+				}
+			} else {
+				return errors.Wrapf(err, "radosgw-admin user info failed with code %d and output %q", code, output)
 			}
+		} else {
+			logger.Debugf("realm system user %q has already been created", uid)
 		}
+	}
 
-		zoneID, err := decodeID(output)
-		if err != nil {
-			return errors.Wrap(err, "failed to parse zone id")
-		}
+	return nil
+}
+
+func setMultisite(context *Context, spec cephv1.ObjectStoreSpec, serviceIP, realmName, zoneGroupName, zoneName, namespace string) error {
+	realmArg := fmt.Sprintf("--rgw-realm=%s", realmName)
+	zoneGroupArg := fmt.Sprintf("--rgw-zonegroup=%s", zoneGroupName)
+	zoneArg := fmt.Sprintf("--rgw-zone=%s", zoneName)
 
-		logger.Infof("RGW: realm=%s, zonegroup=%s, zone=%s", realmID, zoneGroupID, zoneID)
+	serviceEndpoint := fmt.Sprintf("http://%s:%d", serviceIP, spec.Gateway.Port)
```

Comment:
@travisn: I saw @leseb previous PR (https://github.com/rook/rook/pull/5774/files#diff-7e97b37c53b4739ef8382f7b481ab8dcR288) that `Context` has an `Endpoint`. I tried setting `serviceEndpoint` to `context.Endpoint` but got an empty string. Is it being set somewhere else? I'd prefer this way of setting this variable instead of creating it myself.